### PR TITLE
[wip] storage,sql: push cfetcher into pebble's mvcc scanner

### PR DESCRIPTION
**Pull request by jordanlewis**
This commit prototypes injecting the cfetcher state machine into the
mvcc scanner. It adds a new scan format that requests bytes to be
returned in the arrow columnar batch format. Under the hood, the
MVCCScan implementation gets a handle onto a cfetcher, and injects the
pebble iterator into the cfetcher's "Next KV" interface, so that the
cfetcher, when invoked, directly copies keys and values out of the
iterator. Then, the TableReaderSpec at the sql level is implemented by
deserializing the arrow batch directly into a coldata.batch, without
having to do any extra processing.

### CODE REVIEW

File: pkg/roachpb/api.proto [hunk 1]
```diff
@@ -401,6 +405,22 @@ message ScanRequest {
   // keys returned by the request, not a single range lock over the entire span
   // scanned by the request.
   kv.kvserver.concurrency.lock.Strength key_locking = 5;
+
+  // This "any type" field is a serialized TableReaderSpec that can be used
+  // to initialize a columnar scan for MVCC. We use this generic type to resolve
+  // some painful circular dependency issues.
+  google.protobuf.Any scan_spec = 6;
```
Comment (petermattis) [inline pkg/roachpb/api.proto:412]: I suspect we'd want to do something more principled here if this is productionized.

File: pkg/roachpb/api.proto [hunk 2]
```diff
@@ -401,6 +405,22 @@ message ScanRequest {
...
+
+  bool projection = 7;
+
+  // The output columns describe a projection on the internal set of columns;
+  // only the columns in this list will be emitted.
+  //
+  // Can only be set if projection is true. Cannot be set at the same time with
+  // render expressions.
```
Comment (petermattis) [inline pkg/roachpb/api.proto:420]: What is "render expressions" referring to here?

**Review by petermattis — COMMENTED** [review]
Neat to see that this is possible, though I'm not terribly fond of having KV call back into the SQL code in this manner. Is there any hope of disentangling `ColFetcher` from SQL to a greater degree?

`DistSender` splits apart `ScanRequests` on range boundaries and then joins them back together. See `ScanResponse.combine`. Hope that combining responses isn't too difficult with the `COL_BATCH` format.

**Comment by jordanlewis** [comment]
For anyone looking at this later, note that this PR still doesn't reap the performance gains we were expecting. The reason is that, even though storage is writing directly to `coldata.Batch` and avoiding the extra copy from keys to `coldata.Batch` later in the pipe, we're forced to convert the batch into arrow format and then into bytes, and then from bytes back to arrow back to `coldata.Batch`. This is expensive.

There’s 2 directions I want to try to going to fix this. One is making the conversion back and forth from coldata.Batch to arrow batch to bytes more efficient. The other is making a fast path that gets enabled when there is no network hop. The fast path would skip serialization and return an opaque pointer to the coldata.Batch.

**Comment by jordanlewis** [comment]
@sumeerbhola since this came up earlier, I'm very curious to know what you think of it. It's obviously badly out of date. But I think it's an interesting direction to think about, and we'd have to solve problems of this nature if we ever wanted to make a more sql-aware semicolumnar format in Pebble one day.

**Comment by sumeerbhola** [comment]
> But I think it's an interesting direction to think about, and we'd have to solve problems of this nature if we ever wanted to make a more sql-aware semicolumnar format in Pebble one day.

Glad you reminded me -- looking at this PR has been on my TODO list for a few months now.
This is cool stuff.

I can understand why this takes the approach of mostly hiding what is happening from the storage layer internals, since the only thing changing is the output format, and injecting a different implementation suffices for that. Alternatively, we could place the `coldata.Batch` writing code in a place where storage can depend on it directly. 

The reason I am interested in a more transparent approach is that when we have a block-columnar format in Pebble we will need to know about some aspects of the schema in order to lay things out when writing, and use it at read time to decide what parts of a file to read (due to projections) and what filters to apply (including the case where there are column dictionaries at the file level). I think we need to first do some bottom up work in file format design and prototyping to prove out the benefits (write throughput can suffer if we are not very careful, and making sure compactions are very efficient is important). I don't think we are going to start on it very soon.

So if you can show significant performance improvements here by optimizing for the local path, say using the bit in the `AdmissionHeader`, it would be worthwhile doing something like this as a first step (though probably with making the storage package fully aware of what it is doing).

**Comment by jordanlewis** [comment]
> Alternatively, we could place the coldata.Batch writing code in a place where storage can depend on it directly.

I think this would be a better approach architecturally, but I believe it will be very far from straightforward. The encode/decode path from KV to SQL (both `tree.Datum` or `coldata.Batch`) is extraordinarily complex. Either we would have to import all of `TableDescriptor` and its associated stuff into storage (this seems like it would make a mess out of our abstraction layers), or we would have to come up with another encoding method that wasn't as SQL specific that SQL could use as a target instead of the current system (I'm not sure what this would look like, and not sure if it would materially save us any complexity).

Here is a relatively complete spec of the SQL encoding: https://github.com/cockroachdb/cockroach/blob/master/docs/tech-notes/encoding.md. It is missing the per-datatype encoding methods, which are to my knowledge defined only in the code here: https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/rowenc/column_type_encoding.go (note that there are 3 separate encodings per type: key ascending, key descending, and value.).

**Comment by jordanlewis** [comment]
Latest changes:

I implemented a fast-path that skips serializing the scan's ProcessorSpec, by passing an opaque `interface{}` in the protobuf that has no serialization methods. We could theoretically only use this path for local requests, and serialize the ProcessorSpec normally for remote requests (this part is not yet implemented).

The results are potentially promising - finally, the `plan_direct_scan` option causes local scans over TPCH data to perform as good if not slightly (~5%) better than without the option. It makes sense - we've managed to remove one copy from the system.

The benchmarking methodology I used to determine that number was far from good - just running on my local machine - I would like to do some more tests for this soon.

Even if we don't end up merging this, I find it encouraging that we can potentially squeeze out some performance wins by eliminating one of the extra copies (instead of writing to KV batches then writing to col batches, just write directly to col batches).

**Comment by jordanlewis** [comment]
This is superceded by @yuzefovich's ongoing work. Woohoo!

**Pull request closed without merging**